Build optimizations

gonzalezjo opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Build/Testing/Packaging Improvement


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Addition of new options to CMakeLists.txt to try to improve build performance.


Detailed description / Documentation draft:


# Build system optimizations

I'd like to first begin this by offering a tl;dr: Clobber/fresh build times and the final size of the ClickHouse binary can be reduced dramatically based on fairly mild tradeoffs. Incremental build performance may be similarly improved, though in some cases, may also regress. 


On systems with slow IO, we can especially speed up builds times. In full ClickHouse builds with many threads on a ramdisk, I've seen debug build times drop by over 30%. Gains can vary depending on system configuration and build type. Incremental builds are notably faster with these changes when not editing code in the Functions folder.


Performance-wise, the gains from this PR depend a lot on the hardware and options that you build with, so I'm wary of offering specifics. When building with 16 threads, I see a ~19% improvement in build times. With 48 threads, I see a ~9% improvement. All tests were performed with a full build, on a ramdisk, in debug. Gains are expected to be larger on slower storage mediums.


## Research


There are a number of performance constraints in builds. These issues range in scale and size.


Several targets and sets of targets stand out in taking a substantial amount of time to build: the clickhouse_functions targets, the LLVM targets, and those introduced by large libraries like poco and abseil. These are targets that are largely unavoidable; that is, you are never going to compile ClickHouse without compiling clickhouse_functions. 


Post-build, even with lld, incremental builds are typically dominated by linking times, except in circumstances such as those when working in and around clickhouse_functions.


### Research: Our code

The most relevant target for source-level optimization is clickhouse_functions, as it is the only one within our control.


In clickhouse_functions, based on clang's -ftime-trace, we spend just under a third of the time in clickhouse_functions doing template instantiations; breaking that down, a third of that is just spent in `make_default_wrapper`, the name of a lambda in FunctionConversion.h--to be honest, this seems unreasonable to me, and I question whether -ftime-trace is giving reasonable outputs. 


While a few headers make up a fairly substantial percent of build time, the headers themselves are not to blame; rather, its just typical template driven code explosion, and the most-used templates in source files are to blame. This is, of course, expected.


As a consequence of code explosion, however, we get extremely large static libraries. On a hard-drive, any efforts to reduce code bloat can substantially reduce build times by reducing IO.


Fixes for our code are outside of the scope of this branch, so for the sake of brevity, I will leave it at that. I will note that side effects of solutions 1. and 2. partially mitigate some code issues.  


### Research: Third-party dependencies


LLVM spends an obnoxious enormous amount of time running TableGen on fresh debug builds. There's *very* little that can be done about it. We can set `LLVM_OPTIMIZED_TABLEGEN` to make it run faster in debug builds. However, this is is broken in our current version of LLVM, because LLVM assumes that you're building it as a standalone project. This is only slightly fixable if we merge LLVM's D78913, which *partly* addresses this problem. However, even with that patch, building the LLVM subproject with LLVM_OPTIMIZED_TABLEGEN still takes an extremely long time.


LLVM, alongside all of our other dependencies, spends its remaining time linking and compiling. This ends up being a pretty substantial amount of our build-time. This is a reason why it may be worth considering making dynamic linking the default for debug builds.


### Research: Optional targets


We build a vast number of optional targets like Cassandra, tests, etc. While these targets may be useful, they are not always necessary, and their builds can be disabled entirely if needed using existing options. However, in the case that they are needed, it helps somewhat to speed up their builds.


### Possible solutions


1. Unity builds: these substantially improve fresh build times by combining source files into a single compilation unit. This creates smaller and faster-linking static libraries (as the optimizer has more to work with) and reduces redundant computation. (E.g. parsing headers over and over.)


I have not heard any critique of this yet, which I take to be a good sign!


2. Disabling linker optimizations: this seems to *marginally* improves re-linking times for me when I use lld. There are no notable downsides that I am aware of.


Based on feedback from @traceon, this would make more sense as the default for debug builds. The PR will be updated to reflect that.


3. Dynamic linking: dynamically linking as much as possible helps. We mostly support this already. There could be value in making dynamic linking the default for debug builds; however, the performance regressions may show up on CI if any tests are ran in debug, and I don't want to risk breaking any compatibility. If someone wants this as part of the PR, I think that it'd be a trivial change to add.


@traceon has noted that this would be a concern for release build performance and--in the case of out of tree builds--debuggability. I'd like to emphasize that this would be exclusively for in-tree debug builds. But I don't want to die on this hill. If there's no interest, then I do not see a need for it.


4. Code changes using e.g. concepts. This risks compatibility issues with the Yandex build machines, but feature tests and abstraction may make that a non-issue.


@myrrc has noted that this is likely to increase maintenance burden and may be tricky due to compiler compatibility requirements. Concepts specifically are therefore impractical. Other changes are still on the table.


5. Building fewer optional targets / doing a minimal build by default. This would risk breaking compatibility with any systems that automatically build clickhouse and expect those optional targets to be built and used, so we don't do that in this PR. In any case, we already have a build script to do this, so I don't think this is useful except to make ClickHouse build times less intimidating to newer contributors who haven't yet discovered the build scripts.


Based on @traceon's comments, this seems to violate the philosophy of ClickHouse builds, and so I believe that this can be ruled out. 


6. Produce less and link in less debug information: this is possible. Users may not always want all debug information. We can split DWARF files and strip some debug information to try to reduce IO costs. On slow disks, this can be a win.


The better approach, noted by @myrrc, is likely to be to provide #14657-esque knobs for tuning debug information output while still providing sensible defaults.


7. Compress debug information to further reduce IO costs: Even when building to a pretty slow (magnetic/spinning rust) hard drive, this actually--and reproducibly--slowed down build times to a statistically significant degree by making linking slower. I suspect that this may be exaggerated because of filesystem caching that I haven't been able to disable, and that flushing caches between relink tests is insufficient for. Everyone's OS already has a filesystem cache by default, so I felt comfortable removing this from the PR.


Based on my own measurements, this is also ruled out.


### What this PR does


This PR adds support for unity builds for clickhouse_functions related targets and the contrib (dependencies) folder. 


It adds support for splitting DWARF information, and it adds support for removing pubnames. Even with split DWARFs, pubnames get linked in, so these features pair well together.


Finally, the PR adds support for disabling linker optimizations.


These features are introduced as new options. Only the use of unity builds and the disabling of linker optimizations are on by default.


### What this PR doesn't do


Unfortunately, I've struggled immensely with C++ related, and with LLVM build toolchain related attempts to optimize build times. I plan to keep at those, but I want to get this code--which I know works--in. It's not worth keeping this blocked on fairly disparate changes that can be put into a separate PR.

gonzalezjo committed:
Add build optimizations

Add a series of "FASTBUILD" commandline options that can be used to accelerate builds.

gonzalezjo committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into buildopt

gonzalezjo committed:
Misc. CMakeLists.txt changes

Slightly improved output messages.

gonzalezjo committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into buildopt

gonzalezjo committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into buildopt

alexey-milovidov commented on CMakeLists.txt:
```diff
@@ -565,3 +565,139 @@ add_subdirectory (utils)
+
+    # Collect files that trigger C++ ODR violations, so as to exclude them.
+    file(GLOB ODR_VIOLATIONS RELATIVE ${CMAKE_SOURCE_DIR}
+         "src/Functions/toDayOf*.cpp" "src/Functions/normalize*.cpp" "src/Functions/format*.cpp"
```
Maybe just fix these violations?

gonzalezjo committed:
Disable linker optimizations for debug builds not on Darwin

Set -Wl,-O0 when not linking for Darwin. This disables linker optimizations (e.g. string merging) which may improve link times and increase predictability of linker behavior.

traceon commented:
> Performance-wise, the gains from this PR depend a lot on the hardware and options that you build with, so I'm wary of offering specifics. When building with 16 threads, I see a ~19% improvement in build times. With 48 threads, I see a ~9% improvement. All tests were performed with a full build, on a ramdisk, in debug. Gains are expected to be larger on slower storage mediums.

@gonzalezjo Are those 16 and 48 physical, logical, or software threads? Do you have numbers for memory consumption for different combination of flags?

Since, long build times mostly cause noticeable problems during the edit-compile-debug cycles for developers on their machines, which are performed using relatively slow storage and consumer-grade CPU/RAM, and are not clean builds, do you have any numbers of improvements in such environments?

alexey-milovidov commented:
> Compress debug information to further reduce IO costs

I also considered this option but did not enable it because it is not supported by our embedded DWARF interpreter that is used to show line numbers in stack traces (see `Dwarf.h,cpp`).

alexey-milovidov commented on CMakeLists.txt:
```diff
@@ -195,6 +195,13 @@ endif ()
 option(ADD_GDB_INDEX_FOR_GOLD "Add .gdb-index to resulting binaries for gold linker.")
 
 if (NOT CMAKE_BUILD_TYPE_UC STREQUAL "RELEASE")
```
But we never use Release, and use RelWithDebInfo for releases instead.

gonzalezjo committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into buildopt

gonzalezjo committed:
Fix ODR violations in src/Functions

Enables unity building all code in src/Functions.

gonzalezjo committed:
Address improper setting of CMAKE_EXE_LINKER_FLAGS

Conforms with expected behavior now.

gonzalezjo commented:
> > Performance-wise, the gains from this PR depend a lot on the hardware and options that you build with, so I'm wary of offering specifics. When building with 16 threads, I see a ~19% improvement in build times. With 48 threads, I see a ~9% improvement. All tests were performed with a full build, on a ramdisk, in debug. Gains are expected to be larger on slower storage mediums.
> 
> @gonzalezjo Are those 16 and 48 physical, logical, or software threads? Do you have numbers for memory consumption for different combination of flags?
> 
> Since, long build times mostly cause noticeable problems during the edit-compile-debug cycles for developers on their machines, which are performed using relatively slow storage and consumer-grade CPU/RAM, and are not clean builds, do you have any numbers of improvements in such environments?

48 logical threads running on 24 hyperthreaded cores. I do not have absolute numbers on memory consumption, since it varies greatly on how many cores CMake decides to assign to linking, what stage of the build I'm in, etc. But if there are configurations you're interested in, I can provide such numbers.


Memory usage doesn't seem notably increased, and is probably actually decreased when linking, when using Unity builds.


I'll be running a few builds queued, with ccache off, to see the differences in build times when using 4 cores.

alexey-milovidov commented on src/Functions/normalizeQuery.cpp:
```diff
@@ -14,7 +14,7 @@ namespace ErrorCodes
 }
 
-namespace
+namespace NormalizeQuery
```
After this change symbols become exported. Maybe leave anon namespace and rename symbols? (or make two nested namespaces but it will confuse reader).

gonzalezjo replied:
> After this change symbols become exported. Maybe leave anon namespace and rename symbols? (or make two nested namespaces but it will confuse reader).

Renaming is definitely a better choice. Thanks.

myrrc commented on CMakeLists.txt:
```diff
@@ -565,3 +575,121 @@ add_subdirectory (utils)
+# Various build speed optimizations.
+option(FASTBUILD_CONTRIB_UNITY "Use unity builds on the contrib/ folder. (No-op on CMake < 3.16.0)" ON)
+option(FASTBUILD_FUNCTIONS_UNITY "Use unity builds on the src/Functions folder. (No-op on CMake < 3.19.0)" ON)
+option(FASTBUILD_NO_PUBNAMES "Do not emit DWARF pubnames when using Clang. Slows GDB symbol loading. Reduces IO." OFF)
+option(FASTBUILD_SPLIT_DWARF "Build with split DWARF files. Makes debugging symbols non-portable. Reduces IO." OFF)
```
Hi!
As you may know, CH has an automatic [options generator page](https://github.com/ClickHouse/ClickHouse/blob/master/docs/tools/cmake_in_clickhouse_generator.py) (builds to [this](https://clickhouse.tech/docs/en/development/cmake-in-clickhouse/)) which supports additional information that should be added as comments right before the options line, e.g.

```cmake
# Additional option data as restrictions (e.g. clang only) and other useful data.
option(OPTION "Description." OFF)
```

Some options already follow this guideline, so it would be nice if you moved the options descriptions' limitations and so into the comments.

myrrc replied:
For example, the option FASTBUILD_CONTRIB_UNITY actually sets the unity builds for particular targets, so it's better to name them in the commens so the user would not need to look into the source code.

myrrc commented on CMakeLists.txt:
```diff
@@ -296,6 +296,16 @@ set (CMAKE_C_FLAGS                       "${CMAKE_C_FLAGS} ${COMPILER_FLAGS} ${C
+# By setting the options, builds (and in particular, linking stages) may be marginally faster.
+# Additionally, the linker behavior will fall closer in line with what the user would expect.
+# For releases, we allow the linker optimizer to do whatever it can, so options are left as-is.
+if (CMAKE_BUILD_TYPE_UC STREQUAL "DEBUG")
+    if (NOT OS_DARWIN)
+        set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,-O0")
+    endif()
+endif()
+
```
I would consider adding it as a separate flag e.g. DISABLE_LINKER_OPTIMISATIONS and adding the log message as the Debug build is extremely common among developers (and this PR implicitly changes the default behaviour with no notification), e.g.

```cmake
if (DISABLE_LINKER_OPTIMIZATIONS)
 # do stuff
 # warning so users could grep and see the non-default (non-expected) behaviour
 message(WARNING "Disabling linker optimizations")
else()
 message(STATUS "Enabling linker optimizations")
endif()
```

gonzalezjo replied:
That is very nearly what the PR did, before it was implied that this should be the default for debug builds. Should I be changing it back?

myrrc replied:
It's your choice whether to turn this on or off by default, I just say that this option should also be added as a flag (so one could switch it if needed).

myrrc commented on CMakeLists.txt:
```diff
@@ -565,3 +575,121 @@ add_subdirectory (utils)
+
+# Check Unity build compatibility.
+if ((FASTBUILD_CONTRIB_UNITY OR FASTBUILD_FUNCTIONS_UNITY) AND CMAKE_VERSION VERSION_LESS "3.16.0")
+    message("Disabling Unity builds. (Needs CMake >= 3.16.0)")
+    set(FASTBUILD_CONTRIB_UNITY OFF CACHE BOOL INTERNAL FORCE)
```
I would consider changing this message's status to "warning" (and also statuses of all CMake-related optimizations messages) because, unlike ccache and other external tools, CMake is mandatory to this project and users should be aware that the old versions would provide worse build times / binary size.

alexey-milovidov commented:
@gonzalezjo Need to fix the "Fast test", see the logs in "Details".

gonzalezjo commented:
# A Look at the Data on Realistic Hardware


### Challenges 


Returning to this after a brutal past day and a half of attempting to collect data. Attempting to benchmark this on slow storage was a bad idea: I only had cheap USB drives, and they slow down a lot after a couple of builds because of the cheap NAND and controllers. In fact, my first attempt to test this seemingly killed a drive. It won't let me write to it anymore. I went to BestBuy, got two more flash drives, and got really weird results where things started fast and performance quickly got awful. I'm not sure what the best way to emulate slow IO is at this point, so I gave up, and decided to do my benchmarks on an SSD. I'm pretty sure this was somewhat responsible for my counter-intuitive results re: compressed debug info, but I don't think it's worth buying a harddrive to find out; as was noted, compressed debug info breaks some tooling anyway.


I had previously tried to do VM measurements through AWS, but (1) performance ended not being nearly consistent enough to get useful data, and (2) the builds take so long, and I'd need so many samples to get statistically significant measurements, that I just can't justify the AWS costs.


### Methods

So, I devoted my desktop to running tests with four cores, ccache (4.0) enabled, and ninja on a SATA SSD. CPU governor was locked on performance. Before each build, ccache -C was automatically run. Besides the script running all of these builds (each build was run independent), systemd, the windowing manager, etc., nothing else was running.

Because I'm now deeply paranoid of flash controllers and performance inconsistency, after each build, the build folders were automatically deleted and fstrim was automatically ran. After it ran, I slept for one minute (in case the SSD needs to run any garbage collection) and started the next build afterward.

The source files were kept in a tmpfs so that reads to source files wouldn't be affected by OS caching, flash quirks, etc.

These are the results that I have. Note that the defaults are just unity building contrib and clickhouse_functions. "All on" would be additionally stripping pubnames and using split-DWARF.


### Results

| Build type, build optimizations | Time (H:MM:SS) |
|---------------------------------|----------------|
| Debug, all on                   | 1:35:26.52     |
| Debug, defaults                 | 1:16:34.06     |
| Debug, all off                  | 1:33:11.58     |
| Debug, contrib unity build      | 1:24:53.85     |
| Debug, Functions unity build    | 1:26:15.66     |
| Release w/debug info, all on    | 1:46:22.53     |
| Release w/debug info, defaults  | 1:41:14.41     |
| Release w/debug info, all off   | 2:02:31.12     |


### Discussion

I was originally going to run just all on, defaults, and all off benchmarks. I decided to measure the impact of contrib unity builds and functions unity builds separately for debug, just to capture the individual difference that they make. While I can say, based on previous attempts at measuring performance, that contrib unity builds are more helpful to build times, the difference in this data is probably within a 95% CI. I did these benchmarks on debug since debug builds finish faster, but if you want me to run it on release, I can. Just expect similar numbers.


The picture is fairly clear: Unity builds are a pretty big win and stripping pubnames/using split-DWARF can actually lead to build time regressions. Hence, they're disabled by default.


The question worth asking is: why strip pubnames, and why use split-DWARF at all? The benefits aren't compelling in the table I've just posted, so I've written a bit about my motivations below. If you're satisfied with leaving it in, defaulted to off, then you don't have to read below. 


### Why split DWARF and why no pubnames? 

(tl;dr: it's all about IO and incremental builds)

On harddrives (particularly the slower, more energy efficient ones that you see in laptops) and very slow flash storage (like the USB drives I had that ClickHouse builds have massacred :stuck_out_tongue:) you're in a good spot if your drive can do consistent 100 MB/s reads and writes spanning over ~64 MB (or however big the internal cache is...most are 64MB.) In fact, laptop harddrives seem to struggle to do 80 MB/s sequential. ([example](https://hdd.userbenchmark.com/SpeedTest/130/ST1000LM024-HN-M101MBB))

Modern SSDs usually hit a sustained ~500 MB/s on SATA, easily, and on PCIE4.0, can hit [over 7 GB/s](https://www.tomshardware.com/news/intel-debuts-worlds-fastest-ssd-the-pcie-40-optane-ssd-p5800x). So on modern SSDs, trying to reduce IO is often pointless, and on slower drives, that's not really the case. Every gigabyte in writes can be 12.5 seconds or more of IO. Over a clean build, on harddrives, there's therefore the potential to save pretty large amounts of IO by trading off CPU time for writes--which is exactly what gsplit-dwarf does.

Nowhere is that more evident, in fact, than in the type of builds that people do the most: incremental builds. On a harddrive, incremental builds still involve relinking. By splitting debug info off into files that _do not change_ (outside of the few files that the programmer has modified and their dependencies), you no longer find writings tons and gigabytes worth of debug information during the linker stage when doing incremental builds. This means the IO cost of incremental release w/debug info and debug builds can be reduced fairly substantially. This actually has the added benefit of _reducing_ memory usage during builds. When not using LLD (which is so fast that split-DWARF related bookkeeping makes it slower) it might even speed clean builds up. On harddrives, incremental builds should be much faster. On a fast SSD, incremental builds are marginally slower for me, since LLD's split-DWARF implementation isn't that fast and I'm far more CPU constrained than IO constrained. 


For an article on split-DWARF that's far more insightful than anything I can provide, you can [see here](https://www.productive-cpp.com/improving-cpp-builds-with-split-dwarf/). It found that incremental build times were halved by using split DWARF in LLVM. On typical harddrives and slow flash, I would expect similar results for ClickHouse incremental builds.

azat commented:
@gonzalezjo Am I understand right that `Debug, all off` != upstream, since it will include `-Wl,-O0`?

gonzalezjo committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into buildopt

gonzalezjo committed:
CMake compatibility + options + documentation improvements

Improved support for older CMake versions and options now seem to do what was requested. Documentation comments added.

gonzalezjo committed:
Attempt to fix issues with Fast Test

Seems this option breaks contrib unity builds, which makes sense.

gonzalezjo committed:
Style fixes for normalize*.cpp

Use names for uniqueness instead of namespaces as a way to address ODR violations.

gonzalezjo commented:
> @gonzalezjo Am I understand right that `Debug, all off` != upstream, since it will include `-Wl,-O0`?

You are correct.

gonzalezjo committed:
Attempt at style fixes

Unsure why builds still fail.

gonzalezjo committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into buildopt

gonzalezjo committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into buildopt

gonzalezjo committed:
Fix build errors when using g++

g++'s -Wsubobject-linkage treats included .cpp files as included headers. While this isn't strictly wrong or bad, and while it's the behavior you would expect, it is *not* the behavior that you want when using unity builds.
This patch disables -Wsubobject linkage for clickhouse_functions when using unity builds and gcc.

gonzalezjo committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into buildopt

gonzalezjo committed:
Fix build errors with ubsan

Change how we disable LLVM targets and Update comments accordingly. This fixes ubsan errors.

gonzalezjo committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into buildopt

gonzalezjo committed:
Bugfix after merging master

Fixes build error by accessing a function through a  namespace.

gonzalezjo committed:
Attempt to fix thread sanitizer issue

libcxxabi's cxa_guard.cpp causes issues in unity builds because it uses symbols shared with sanitizers.
Disable unity builds for libcxxabi when sanitizers are enabled.

gonzalezjo committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into buildopt

gonzalezjo committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into buildopt

gonzalezjo committed:
Fix merge mistakes

This took way too long to catch.

gonzalezjo committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into buildopt

gonzalezjo committed:
Attempt to fix a pvs_check failure

Consolidate LLVM_OVERRIDE_LIST and a set_target_properties call into one big OVERRIDE_LIST.

gonzalezjo committed:
Refactor CMakeLists.txt

Attempt to avoid issues that show up in tests by just doing things differently.

gonzalezjo committed:
Refactor CMakeLists.txt

Attempt to avoid issues that show up in tests by just doing things differently.

gonzalezjo committed:
Move RocksDB to a lower batch size

Apparently it does NOT support all batch sizes.

gonzalezjo committed:
Don't unity-build rocksdb

There is absolutely no reason why this should fix anything, but it's not failing on others' PRs, so I have to assume that this is causing it?
In fact, this issue should always occur on ARM according to my intuition, but I guess my intuition is violated for some reason.

gonzalezjo committed:
Force CI run

Disregard this :)

gonzalezjo committed:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into buildopt

danlark1 reviewed (commented):
First of all, thank you.

Just a small note: it is much more useful for us in general to optimize the CI builds as they appear to be the most consuming, the developers don't always run all tests before submitting to check everything is fine. It would be consuming and it's usually better to test something locally, send a PR to make sure everything works fine and only them submit. The most valuable time is the feedback loop between the CI and the developer, optimizing 2x time of the bottleneck would likely save 20-30 minutes of it which is awesome. Once the developers check out ClickHouse, it's unavoidable to build everything and incremental times are much more valuable, however, in CI, the building is done constantly almost each second with different workloads and low level patches. Most of the developers who build ClickHouse with low level changes already have pretty strong hardware to do so. So optimizations of build times are better be for strong hardware rather than for "real".

To me the optimizations of 10-15% looks kinda small given my experience with unity builds, I guess this is because more work can be involved: there are many other targets of clickhouse which also can benefit of optimizing the build times

alexey-milovidov commented:
Any idea why Parquet test has failed?

alexey-milovidov commented:
At least all builds have passed. This is great.

alexey-milovidov commented:
@gonzalezjo This PR has a chance.

alexey-milovidov commented:
Recently we have found an issue with the usage of `ccache` in our CI.
After it was fixed most of the builds in CI becomes very fast (as they are started to be incremental).

For example, the binary build in master is only 3 minutes (the last row):
https://clickhouse-builds.s3.yandex.net/0/ca03670e60cfec19c84e1dfb683e74a0eb3ae65c/clickhouse_build_check/report.html

alexey-milovidov commented:
So, this PR does two things:
1. Unity builds.
2. Faster linking.

Linking runs in several seconds if not IO bound. If it's IO bound it runs in the same time as needed to read and write several gigabytes on disk. So, it does not give much to optimize linking stage.

Unity builds makes sense but:
\- the amount of complexity may not pay itself;
\- it requires renaming symbols in anonymous namespaces that is unnatural;
\- due to increased memory usage it may not be applicable on developers machines;
\- we don't need it in our CI due to the usage of ccache;
\- this PR is unfinished (tests are failing).
